### ISSUE TITLE
iOS 10 crash - Layout Attributes

### ISSUE THREAD

**Issue by zachlucas** [CSStickyHeaderFlowLayout/CSStickyHeaderFlowLayout#144]
Getting this crash when scrolling up in a UICollectionView section that has a sticky header:

```
layout attributes for supplementary item at index path (<NSIndexPath: 0xc000000000000116> {length = 2, path = 1 - 0}) changed from <CSStickyHeaderFlowLayout: 0x7fe26a858880> indexPath: {1, 0} zIndex: 1024 valid: YES kind: UICollectionElementKindSectionHeader to <CSStickyHeaderFlowLayout: 0x7fe26a85dd80> indexPath: {1, 0} zIndex: 1024 valid: YES kind: UICollectionElementKindSectionHeader without invalidating the layout
```

It seems like the header is switching incorrectly?

**Comment by GabeJacobs** [comment]
Getting this as well!

**Comment by zachlucas** [comment]
@jamztang any idea on this one?  Seems to happen when scrolling back up.

**Comment by GabeJacobs** [comment]
@zachlucas however this guy does it seems to avoid the crash https://github.com/Produkt/PDKTStickySectionHeadersCollectionViewLayout

**Comment by kevinnguy** [comment]
I believe this is happening due to `UICollectionElementKindSectionHeader` interfering with `CSStickyHeaderParallaxHeader` in your collection view. It might be some priority layout issue.

**Comment by zachlucas** [comment]
@GabeJacobs thanks, I'm also seeing no crash with `PDKTStickySectionHeadersCollectionViewLayout`

**Comment by kevinnguy** [comment]
I was able to avoid the crash by commenting out:
```
if (!CGSizeEqualToSize(CGSizeZero, header.frame.size)) {
    [allItems addObject:header];
}
```
Based on the comment for this implementation, it seems like the layout would invalidate if this wasn't implemented but my app was able to run without it, even with many section headers on view.

**Comment by fitch** [comment]
Commenting out

**Comment by chrene** [comment]
@zachlucas can you create a sample project that crashes so I can investigate the issue?

**Comment by eJamesLin** [comment]
@chrene the CSStickyHeaderFlowLayoutDemo will crash on iOS 10 devices and simulator

**Comment by chrene** [comment]
I see, just updated to xcode 8 and running iOS 10 and it does crash. I saw a pull request that disables prefetching, but I'm trying to find a more robust fix. Basically as a workaround you can disable prefetching on your collection view to avoid the crash.

**Comment by chrene** [comment]
I have also written the library in swift and I'm planning on bumping the version in the near future. I don't get the crash from the swift branch, *BUT* be careful about using that in production. It still needs a lot of testing :)

**Comment by chriscohoat** [comment]
Thanks @chrene! I've disabled prefetching in my UICollectionViewController like so:

```
override func viewDidLoad() {
    super.viewDidLoad()
    if #available(iOS 10.0, *) {
        self.collectionView?.isPrefetchingEnabled = false
    } else {
        //Fallback on earlier versions
    }
}
```

The crash is fixed with this update.

**Comment by FutureChen** [comment]
Thanks @chriscohoat!   useful

**Comment by DevangiDesai** [comment]
Thanks i was facing same problem.

**Comment by rsa2000** [comment]
Using the objective c pod 0.2.10 using isPrefetchingEnabled = NO on the collection view fixes the crash , but when scrolling with sticky headers the header cell disappears where the next header cell is visible, could you please fix the bug in the objective c branch

**Comment by k-sathireddy** [comment]
Thank you so much @chriscohoat .. Worked for me.

**Comment by iMark21** [comment]
Thanks @chriscohoat!! Works like a charm!

**Comment by chrene** [comment]
Another note: if you have prefetching enabled I guess you also have to implement the `UICollectionViewDataSourcePrefetching` protocol in your datasource, but that's out of scope for this project. I think there is some nasty bug in the Objective C code somewhere, but I'm not able to find it. I don't see this bug in the swift branch though.

**Comment by eliasbagley** [comment]
I'm seeing this issue too. Is the fix you're recommending to just turn off prefetching? Is there another way?

**Comment by mihirpmehta** [comment]
I am facing the same exception when i try to scroll up my collectionView Programmatically ... but this is only happening when i am running it on **iOS 11 beta5 using Xcode 9 beta 2**

**Comment by mayanktiwari** [comment]
Yes I am getting this crash as well on Xcode 9 with iOS 11 SDK (public release). But only if there is another section header of kind UICollectionElementKindSectionHeader as well along with CSStickyHeaderParallaxHeader

**Comment by mennolaan** [comment]
for me im experiencing the same error as mayanktiwari.  if #available(iOS 10.0, *) {
        self.collectionView?.isPrefetchingEnabled = false
    } else {
        / . ditn fix it for me though
